cpp/src/arrow/array/data.cc

```diff
@@ -304,9 +304,13 @@ void ArraySpan::FillFromScalar(const Scalar& value) {
 
   Type::type type_id = value.type->id();
 
-  // Populate null count and validity bitmap (only for non-union/null types)
-  this->null_count = value.is_valid ? 0 : 1;
-  if (!is_union(type_id) && type_id != Type::NA) {
+  // Populate null count and validity bitmap (only for non-union/ree/null types)
```

Comment:
this comment can now be removed as the `if` checks wrapping the `null_count` assignments document the context of each assignment